monai/utils/module.py

```diff
@@ -193,7 +211,127 @@ def load_submodules(basemod, load_all: bool = True, exclude_pattern: str = "(.*[
     return submodules, err_mod
 
 
+class ComponentScanner:
+    """
+    Scan all the available classes and functions in the specified packages and modules.
+    Map the all the names and the module names in a table.
+
+    Args:
+        pkgs: the expected packages to scan modules and parse component names in the config.
+        modules: the expected modules in the packages to scan for all the components.
+            for example, to parser "LoadImage" in config, `pkgs` can be ["monai"], `modules` can be ["transforms"].
+        excludes: if any string of the `excludes` exists in the full module name, don't import this module.
+
+    """
+
+    def __init__(
+        self, pkgs: Sequence[str], modules: Optional[Sequence[str]] = None, excludes: Optional[Sequence[str]] = None
```

Comment:
had a quick test of scanning 'itk' and 'torchvision' but it doesn't work... I still feel this is importing is heavy and do we need this if the config is lazily instantiated?
```py
>>> from monai.utils.module import ComponentScanner
ComponentScanner>>> ComponentScanner(["itk"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Documents/MONAI/monai/utils/module.py", line 233, in __init__
    self._components_table = self._create_table()
  File "/Users/Documents/MONAI/monai/utils/module.py", line 246, in _create_table
    module = import_module(modname)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/itk/ITKAnisotropicSmoothingPython.py", line 13, in <module>
    from . import _ITKAnisotropicSmoothingPython
ImportError: PyCapsule_Import could not import module "_ITKCommonPython"
>>> ComponentScanner(["torchvision"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users//Documents/MONAI/monai/utils/module.py", line 233, in __init__
    self._components_table = self._create_table()
  File "/Users//Documents/MONAI/monai/utils/module.py", line 246, in _create_table
    module = import_module(modname)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 670, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 583, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1043, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
ImportError: dynamic module does not define module export function (PyInit__C)
>>> 
```

Reply:
Hi @wyli ,

Thanks for your review and test.
In the overall example, I plan to let users define the `packages` and `modules` in the `metadata.json`.
For 90% MMARs, only need to import `monai` and `torch` with specific `modules`(transforms, networks, etc.)
If you want to use some `ITK` APIs for spetial logic, you can do it in 2 ways:
(1) use "<path>": "itk.XXX" directly like the Hydra example.
(2) Only define the expected APIs of ITK in the `packages` and `modules` instead of the whole ITK.

Maybe you can help provide a simple python logic with ITK, then I try to convert it to a config and parse?

Thanks.

Reply:
Hi @Nic-Ma my question is not about itk. It's about the API we provide here. We mainly offer Python APIs in this package, if they only work for certain use cases of config files we should make it a private class and validate the inputs. This class recursively imports all attributes in the given packages but the actual config might only use a tiny portion for a given workflow, mostly unnecessarily.

Reply:
Hi @wyli ,

That's a good point, I agree it can be a private class. I didn't expect users to use it directly.
In the whole picture, `ConfigParser` will use it internally:
https://github.com/Project-MONAI/MONAI/pull/3593/files#diff-57ffc6dd5a285d7b6e1fa51795ee9e39e296abe50c04006ae0edf0e979af3fbfR52

Let me move the class out of `utils` and be `private`.

Thanks.

Reply:
Hi @ericspod @wyli ,

Thanks for the discussion.
I have a tradeoff idea: for the first MVP, let's only support to load the `classes` or `functions` from MONAI codebase as qualified `<name>`. For APIs from any other packages, just use `<path>` with the full path.
I believe more than 90% APIs in the JSON config should come from MONAI itself.
We can enhance this part later when we get enough feedback later.
What do you think?

Thanks.

Reply:
Hi @ericspod @wyli ,

I updated this part and the unit tests according to both you guys' comments.
Could you please help take a look?

Thanks in advance.

Reply:
Hi @ericspod , does the `load_submodules` already load all the `classes` and `functions` of MONAI code? Then we can call something like `pydoc.locate("LoadImaged")` directly to get the `LoadImaged` class? Seems it should not work for same name APIs.
If not, I think we still need `ComponentScanner`.

Thanks in advance.

Reply:
It was meant to, if it's still being used in the same way then it should continue to do that. Running `python -c 'import monai,sys;print([m for m in sys.modules.keys() if "monai" in m])'` seems to indicate it does.

Reply:
Hi @ericspod @wyli ,

Thanks for the tips, seems `load_submodules` already loaded all the `modules`. I totally simplified the `ComponentScanner` to leverage `load_submodules` results to scan all the `classes` and `functions` directly with a `excludes` filter. And removed the args for `pkgs` and `modules`.
I think the API is simple and clear now, it only works for MONAI components.
Do you have any other comments or concerns?

Thanks in advance.